[e2e] External sites - update /shopper tests, part 1

veljkho opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

In this PR, we're covering some of the shopper tests against external envs. This is going to be the 1st part and there will be 1-2 more to cover them all. This covers 76 tests across the first 13 test files in /shopper/ folder.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. CI green - to be sure that changes in this PR are not affecting wp-env execution.
2. Locally - checkout this branch and run `pnpm test:e2e:with-env default-pressable`.
In `envs/default-pressable/playwright.config.js`, you can comment on everything except this:
```
'**/shopper/account-email-receiving.spec.js',
'**/shopper/add-to-cart.spec.js',
'**/shopper/cart-block-calculate-shipping.spec.js',
'**/shopper/cart-block-coupons.spec.js',
'**/shopper/cart-block.spec.js',
'**/shopper/cart-calculate-shipping.spec.js',
'**/shopper/cart-checkout-block-calculate-tax.spec.js',
'**/shopper/cart-checkout-calculate-tax.spec.js',
'**/shopper/cart-checkout-coupons.spec.js',
'**/shopper/cart-checkout-restricted-coupons.spec.js',
'**/shopper/cart-redirection.spec.js',
'**/shopper/cart.spec.js',
'**/shopper/checkout-block-coupons.spec.js',
'**/shopper/checkout-block.spec.js',
```
3. In the code, I put comments on why these changes.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

veljkho committed:
Update shopper tests and util to fix external env compatibility

veljkho commented on plugins/woocommerce/tests/e2e-pw/utils/editor.js:
```diff
@@ -45,6 +45,7 @@ const goToPageEditor = async ( { page } ) => {
 	);
 	await page.goto( 'wp-admin/post-new.php?post_type=page' );
 	await disableWelcomeModal( { page } );
+	await closeChoosePatternModal( { page } );
```
There are many tests with closing pattern needed. Let's keep this inside the editor util.

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/checkout-block.spec.js:
```diff
@@ -24,7 +24,7 @@ const newAccountEmail = `marge-${ new Date()
 	.getTime()
 	.toString() }@woocommercecoree2etestsuite.com`;
-const newAccountCustomPassword = 'supersecurepassword123';
+const newAccountCustomPassword = 'sup3rS3cur3P4ssw0rd!#123';
```
This is required for WPCOM env, where a strong password is required.

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/cart-checkout-restricted-coupons.spec.js:
```diff
@@ -34,6 +34,27 @@ test.describe(
 				version: 'wc/v3',
 			} );
 
+			// make sure the store address is US
+			await api.post( 'settings/general/batch', {
+				update: [
+					{
+						id: 'woocommerce_store_address',
+						value: 'addr 1',
+					},
+					{
+						id: 'woocommerce_store_city',
+						value: 'San Francisco',
+					},
+					{
+						id: 'woocommerce_default_country',
+						value: 'US:CA',
+					},
+					{
+						id: 'woocommerce_store_postcode',
+						value: '94107',
+					},
+				],
+			} );
```
This is to ensure the store is always US. I experienced at first Afghanistan, might be some leftover from the other tests. Let's keep this for the test to ensure US address.

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/account-email-receiving.spec.js:
```diff
@@ -240,10 +240,10 @@ test.describe(
-						.getByText( 'Password reset email has been sent' )
-						.count()
-				).toBeGreaterThan( 0 );
+					page
+						.getByRole( 'alert' )
+						.getByText( 'Password reset email has been sent.' )
+				).toBeVisible();
```
This was failing on the WPCOM environment. Anyway, updating it to look for an alert rather than verifying the >0 number is a much more convenient assertion.

Stojdza reviewed (commented):
@veljko, when I ran these tests locally, only 55 out of 76 had passed. Did you have a similar experience? 

It was better on CI, [8 shopper tests had failed ](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-ab62a2d/default-pressable-core-e2e/index.html#suites).

veljkho commented:
> @veljko, when I ran these tests locally, only 55 out of 76 had passed. Did you have a similar experience?
> 
> It was better on CI, [8 shopper tests had failed ](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-ab62a2d/default-pressable-core-e2e/index.html#suites).

@Stojdza I'm checking it now and will let you know.
There might be some leftover data conflict from merchant tests, as they also failed.

veljkho commented:
@Stojdza In my latest testing from the local against clean Pressable environment (I had to remove data leftover on site) and after commenting out all other tests and testing only against shopper tests from this PR, the results are that only one test has failed `shopper/cart-checkout-restricted-coupons.spec.js` which I will be looking to fix.

Stojdza commented:
Thanks @veljkho! These are much better results than mine 👍 Are you aiming to fix `shopper/cart-checkout-restricted-coupons.spec.js` as a part of this PR? If not, let's skip them with `@skip-on-default-pressable` and address them in another PR.

veljkho commented:
@Stojdza That is it, the 20 coupons that were left on the test site caused to fail the test that leaned on coupons (the above 1 failing in my testing). After cleaning the test site and rerunning the test, [it finally passed](https://d.pr/i/n1ufSJ).

After the test, I checked the Coupons page on the test site again, and there were no coupons left. We have to check the tests in some of the next PRs to address which tests still leave data on the site after they are run.

veljkho committed:
Update WPCOM config to include shopper tests and skip one from testing

veljkho commented:
@Stojdza I added a new commit including shopper tests from this PR to also WPCOM config as they were passing as well, except only one `shopper/checkout-block.spec.js` failing to start for 400 error reason, no idea why at this moment so skipping it from testing against WPCOM env.

Here are the [testing results against WPCOM-only shopper tests](https://d.pr/i/oR6ckX) from this PR.

veljkho committed:
Update WPCOM config

veljkho committed:
Update shopper shop title deletion test to avoid false positives

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/shopper/shop-title-after-deletion.spec.js:
```diff
@@ -8,6 +8,8 @@ test.describe(
 		test.use( { storageState: process.env.ADMINSTATE } );
 		test.beforeEach( async ( { page } ) => {
 			await page.goto( 'wp-admin/edit.php?post_type=page' );
+			await page.locator( 'input[name="s"]' ).fill( 'Shop' );
+			await page.locator( 'input[value="Search Pages"]' ).click();
```
This is to first search for Shop and then click on it. Sometimes, there are pages left, and the Shop becomes visible on the second page.

adimoldovan replied:
Deleting it via the API would be faster and less likely to cause flakiness.

veljkho replied:
@adimoldovan  The pages? Yes, they should be all auto-deleted, I have no idea why they were left. When I executed all the shopper tests from the local, no pages where left. It seems that sometimes flaky/failing test (shopper or merchant) leaves behind some created pages such as checkout-blocks-12382 etc.

With this workaround to search for the shop page first, we will ensure the shop page will be found before clicking it.

adimoldovan replied:
No, I'm referring to this setup for this test that you updated. It deletes the Shop page, and you have updated it to first search for the page. How about replacing these UI steps with an api call to trash the page?

veljkho replied:
Ah, that's worth updating, as it looks pretty unnecessary.

I'm trying locally to make it work using `wpApi` and I know there is fixtures one, however, even with this current approach of using `beforeAll` & `afterAll` [I don't know why it won't update the page status](https://d.pr/i/SNMvSj).

When I added [simple code to check if my post api works](https://d.pr/i/BIBrj0), it says: Error: `Failed to update shop page status. Status: 401`

The same is even for local `wp-env`, not just external envs. It might be some problem to auth to post... Do you have any advice on how to proceed? @adimoldovan

veljkho replied:
Omg, why I wanted to make the status `draft`, actually I wanted to put the page in the trash. My bad, I will try it that way.

veljkho replied:
It is actually the same even with the `status: 'trash',`.

adimoldovan replied:
@veljkho 
- you get a 401 error, which means you probably set the api client wrong. Why are you not using the `wpApi` fixture instead? It's there for exactly this.
- I don't think you can trash a page by setting it's status. You need to delete it without the `force` argument set to `true`.

adimoldovan replied:
Here's the code with the fixture that works fine locally. Try using this instead of before/after hooks. It doesn't work with the pressable site because it cannot find the page with `slug=shop`. Probably because it was deleted in previous test runs and remained so? You'll need to figure that part out.

```
const { test: baseTest, expect } = require( '../../fixtures/fixtures' );

// test case for bug https://github.com/woocommerce/woocommerce/pull/46429
const test = baseTest.extend( {
	storageState: process.env.ADMINSTATE,
	page: async ( { page, wpApi }, use ) => {
		const response = await wpApi.get( `/wp-json/wp/v2/pages?slug=shop`, {
			data: {
				_fields: [ 'id' ],
			},
		} );

		const pages = await response.json();
		console.log( pages );
		const pageId = pages[ 0 ].id;

		await wpApi.delete( `/wp-json/wp/v2/pages/${ pageId }`, {
			data: {
				force: false,
			},
		} );

		await use( page );

		await wpApi.post( `/wp-json/wp/v2/pages/${ pageId }`, {
			data: {
				status: 'publish',
			},
		} );
	},
} );

test.describe(
	'Check the title of the shop page after the page has been deleted',
	{ tag: [ '@payments', '@services', '@could-be-unit-test' ] },
	() => {
// rest of the code
```

veljkho replied:
Thanks! This is working for me as well. I remember trying to involve fixtures and even `pages?slug=shop` but not to delete it without `force: false`.

I just added a new commit improving the test using `wpApi` fixture and it works like a charm.

veljkho replied:
The shop page on Pressable is in draft status. I remember fixing that already, and it should not occur again.

veljkho replied:
Actually, fixed it in this PR, now with the new code using fixtures:
```
await wpApi.post( `/wp-json/wp/v2/pages/${ pageId }`, {
	data: {
		status: 'publish',
	},
} );
```

Stojdza replied:
> The shop page on Pressable is in draft status.

Do we know how/when this happened?

veljkho replied:
The fix for that is in this PR. Someone executed tests without the fix from this PR of course.

Stojdza replied:
That might be me :)

veljkho replied:
The original test `shop-title-after-deletion.spec.js` didn't clean up after moving the shop page from trash (restoring). Then, I added it via UI to make it published. However with the above `wpApi`, we have it as fixed again.

Commit:
Merge branch 'trunk' into e2e/external-sites/update-shopper-tests-pt1

Stojdza reviewed (commented):
[One shopper test](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-9d33f0e/default-pressable-core-e2e/index.html#suites/817ae9cf6e6243346055dad2a14cd9e6/937d59d69890591/) is failing for Pressable and [one for WPCOM](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-9d33f0e/default-wpcom-core-e2e/index.html#suites/5cab35dd4043f81f4f63ad4ae4456fc6/4e890ccf4a6f1a3a/). Good job overall 👍 

Let's see what will come up from [this](https://github.com/woocommerce/woocommerce/pull/51068/files#r1745150068) discussion, but looks good to me 👍

veljkho commented:
> [One shopper test](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-9d33f0e/default-pressable-core-e2e/index.html#suites/817ae9cf6e6243346055dad2a14cd9e6/937d59d69890591/) is failing for Pressable and [one for WPCOM](https://a8c-woo-test-reports.s3.amazonaws.com/reports/other/on_demand_e2eexternal-sitesupdate-shopper-tests-pt1-9d33f0e/default-wpcom-core-e2e/index.html#suites/5cab35dd4043f81f4f63ad4ae4456fc6/4e890ccf4a6f1a3a/). Good job overall 👍
> 
> Let's see what will come up from [this](https://github.com/woocommerce/woocommerce/pull/51068/files#r1745150068) discussion, but looks good to me 👍

I forgot to add also search for Shop and click for the `afterAll`, as I added only for `beforeAll`.

veljkho committed:
Add search and click steps to afterEach

veljkho committed:
Update selector in test to pass across all envs

veljkho committed:
Merge branch 'trunk' into e2e/external-sites/update-shopper-tests-pt1

veljkho committed:
Update external PW configs for shopper tests

veljkho commented:
@Stojdza thanks for reviewing and reporting the failing tests. I addressed them and now they are fixed, tested across all envs locally. I also updated PW configs for shopper tests. If you think the change is good please approve so I could merge.

The test that require update due to necessary UI steps can be addressed in a separate PR because that is not part of the project but worth updating until we figure out how to do that via wpApi.

adimoldovan commented:
> The test that require update due to necessary UI steps can be addressed in a separate PR because that is not part of the project but worth updating until we figure out how to do that via wpApi.

@veljkho If you decide to skip the api part the test still fails. At one point you need to make another search for the page. I believe the trouble to make it work via the UI is event bigger than figuring out the api part.

veljkho commented:
At which point to make another search for the page? @adimoldovan

adimoldovan commented:
> At which point to make another search for the page? @adimoldovan

I'm not sure. I switched to another branch now, but it was failing for me locally. On the other hand the Shop page is missing from the site now, maybe that was the reason. I gave you a functioning fixture with the API, I still suggest going that route. It's faster, cleaner and less flaky. But you'll need to restore the page first.

veljkho committed:
Improve existing test to use WP API for setup

veljkho commented:
Aha, maybe you didn't have latest commit where I also added search for the page in `afterEach`, so that another search you meant. Generally someone who wrote this test didn't know about using WP API to set it up. We might need to spread knowledge outside Solaris to avoid spending time fixing tests after they were added.

Commit:
Merge branch 'trunk' into e2e/external-sites/update-shopper-tests-pt1

Stojdza reviewed (approved):
Looks good to me 👍

I'm glad we're using API instead of UI now.
